extensions/amp-instagram/1.0/instagram.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {ContainWrapper} from '../../../src/preact/component';
+import {getData} from '../../../src/event-helper';
+import {parseJson} from '../../../src/json';
+import {useMountEffect} from '../../../src/preact/utils';
+import {useRef, useState} from '../../../src/preact';
+
+/**
+ * @param {!InstagramPropsDef} props
+ * @return {PreactDef.Renderable}
+ */
+export function Instagram({shortcode, captioned, style, title, resize}) {
+  if (style == undefined || style['width'] == undefined) {
+    throw new Error(`The width attribute is required.`);
+  }
+  const iframeRef = useRef(null);
+  const [heightStyle, setHeightStyle] = useState(style['height']);
+  const [opacity, setOpacity] = useState(0);
+
+  useMountEffect(() => {
+    /**
+     * @param {Event} event
+     */
+    function handleMessage(event) {
+      if (
+        event.origin != 'https://www.instagram.com' ||
+        event.source != iframeRef.current.contentWindow
+      ) {
+        return;
+      }
+
+      const data = parseJson(getData(event));
+
+      if (data['type'] == 'MEASURE') {
+        const height = data['details']['height'];
+        if (typeof resize === 'function') {
+          resize(height);
+        } else {
+          setHeightStyle(height);
+        }
+        setOpacity(1);
+      }
+    }
+
+    window.addEventListener('message', handleMessage);
+
+    return () => {
+      window.removeEventListener('message', handleMessage);
+    };
+  });
+
+  return (
+    <ContainWrapper
+      className="amp-instagram-container"
+      style={{...style, 'height': heightStyle}}
+      layout
+      size
+      paint
+    >
+      <iframe
+        ref={iframeRef}
+        src={
+          'https://www.instagram.com/p/' +
+          (encodeURIComponent(shortcode) || 'error') +
+          '/embed/' +
+          (captioned ? 'captioned/' : '') +
+          '?cr=1&v=12'
+        }
+        scrolling="no"
+        frameborder="0"
+        allowtransparency
+        title={title}
+        style={{
+          width: '100%',
+          height: '100%',
+          opacity,
```

Comment:
So, there's always some sort of "starting" message from the iframe? If so, we have to switch to `useLayoutEffect` for that, otherwise we risk missing it due to a race condition between an iframe and Preact scheduler.

Reply:
For future reference: 

The sequence is this:
1. Preact inserts the iframe into DOM.
2. Browser schedules iframe load.
3. Preact's `useEffect` is called. It sets up message listener.
4. Iframe is loaded and sends "I'm ready" message of some sort.
5. The message listener setup in (3) shows the iframe.
So, with `useEffect`, we can't guarantee that 3 and 4 will be in this order.
It's very possible that a cached Instagram page will execute (4) before Preact executes (3)
In this case, the step (5) will be missed.

`useLayoutEffect` ensures that the message connection code will be executed in the same task and thus (3) and (4) will be in correct order.

`useEffect` is called "some time" after the rendering is done. "Some time" is up to the framework.`useLayoutEffect`, however, is called immediately after the component is rendered in the same task.
The iframe loading can complete only on a separate task. Thus useLayoutEffect will always be before iframe can message.

This doesn't happen often: `useEffect` will be firing within 50ms or so of rendering, while iframe loading would take at least ~1s.
But (a) the iframe could be cached, in which case it will fire much faster
(b) we don't want to have a race condition if we can avoid it.